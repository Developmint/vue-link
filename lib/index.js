export default {
  functional: true,
  render (h, { data, children, props }) {
    // If no to is set, treat as "external" to remove "link style" and bindings.
    // Quite useful if you want to bind a link dynamically and don't want to have it clicked and styled when not bound
    const isExternal = props => !props.to ||
      props.to.startsWith('http') ||
      props.to.startsWith('//') ||
      props.external

    return isExternal(props)
      ? h('a', {
        ...data,
        attrs: {
          href: props.to || undefined,
          rel: props.rel || undefined,
          target: props.target || (props.newTab ? '_blank' : undefined)
        }
      }, children)
      : h('router-link', data, children)
  }
}
