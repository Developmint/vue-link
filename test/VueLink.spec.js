/* eslint-disable no-console */
import { createLocalVue, mount, RouterLinkStub } from '@vue/test-utils'
import VueRouter from 'vue-router'
import { VueLink, VueLinkAddSlash, VueLinkStripSlash } from '../lib'

const localVue = createLocalVue()
localVue.use(VueRouter)

describe('VueLink', () => {
  describe('router-link', () => {
    it('does link to the correct internal page', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: '/test'
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.contains(RouterLinkStub)).toBe(true)

      const link = wrapper.find(RouterLinkStub)

      expect(link.vm.$props.to).toBe('/test')
    })
  })
  describe('external', () => {
    it('does trigger external on http link', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: 'http://example.com/'
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does trigger external on https link', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: 'https://example.com/'
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does trigger external on // link', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: '//example.com/'
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does trigger external on external prop set', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: '/far-away',
            external: true
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does set rel correctly', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: 'http://example.com',
            rel: 'nofollow'
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does set target on newTab correctly', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: 'http://example.com',
            newTab: true
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does set target on prop correctly', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        context: {
          props: {
            to: 'https://example.com',
            target: '_parent'
          }
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does not bind href if unset', () => {
      const wrapper = mount(VueLink, {
        localVue,
        attachToDocument: true,
        stubs: {
          RouterLink: RouterLinkStub
        },
        slots: {
          default: '<div>Hi</div>'
        }
      })

      expect(wrapper.isVueInstance()).toBe(false)
      expect(wrapper.contains(RouterLinkStub)).toBe(false)

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

describe('VueLinkStripSlash', () => {
  it('strips slashes', () => {
    const wrapper = mount(VueLinkStripSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test/'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test')
  })
  it('strips from link with hash', () => {
    const wrapper = mount(VueLinkStripSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test/'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test')
  })
  it('does not strip anything if no slash present', () => {
    const wrapper = mount(VueLinkStripSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test#abc'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test#abc')
  })
  it('does not strip anything from link with hash without slash', () => {
    const wrapper = mount(VueLinkStripSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test#abc'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test#abc')
  })
})

describe('VueLinkAddSlash', () => {
  it('adds slashes', () => {
    const wrapper = mount(VueLinkAddSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test/')
  })
  it('adds slashes before hash', () => {
    const wrapper = mount(VueLinkAddSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test#abc'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test/#abc')
  })
  it('does not add second slash', () => {
    const wrapper = mount(VueLinkAddSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test/'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test/')
  })
  it('adds second slash before hash', () => {
    const wrapper = mount(VueLinkAddSlash, {
      localVue,
      attachToDocument: true,
      stubs: {
        RouterLink: RouterLinkStub
      },
      context: {
        props: {
          to: '/test/#abc'
        }
      },
      slots: {
        default: '<div>Hi</div>'
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.contains(RouterLinkStub)).toBe(true)

    const link = wrapper.find(RouterLinkStub)

    expect(link.vm.$props.to).toBe('/test/#abc')
  })
})
