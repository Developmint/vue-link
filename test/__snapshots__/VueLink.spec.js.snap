// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueLink external does not bind href if unset 1`] = `
<a rel="noopener">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does set rel correctly 1`] = `
<a href="http://example.com" rel="nofollow">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does set target on newTab correctly 1`] = `
<a href="http://example.com" rel="noopener" target="_blank">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does set target on prop correctly 1`] = `
<a href="https://example.com" rel="noopener" target="_parent">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does trigger external on // link 1`] = `
<a href="//example.com/" rel="noopener">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does trigger external on external prop set 1`] = `
<a href="/far-away" rel="noopener">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does trigger external on http link 1`] = `
<a href="http://example.com/" rel="noopener">
  <div>Hi</div>
</a>
`;

exports[`VueLink external does trigger external on https link 1`] = `
<a href="https://example.com/" rel="noopener">
  <div>Hi</div>
</a>
`;
